File: ai-agent-service/src/main/java/com/enterprise/ai/agent/scan/ScanProjectBlockers.java
package com.enterprise.ai.agent.scan;

import java.util.List;

/**
 * 扫描项目「删除 / 非增量重新扫描」前的外部引用检测结果。
 * <p>
 * 全局 {@code tool_definition} 中 {@code project_id} 指向本扫描项目的 Tool / Skill，
 * 若仍被某 Agent 的 tools / skills 白名单引用，则不允许删除或全量重扫（避免破坏线上 Agent）。
 */
public record ScanProjectBlockers(
        boolean blocked,
        List<String> toolNames,
        List<String> skillNames,
        List<AgentRef> agents
) {
    public static ScanProjectBlockers empty() {
        return new ScanProjectBlockers(false, List.of(), List.of(), List.of());
    }

    public record AgentRef(String id, String name) {
    }
}

File: ai-admin-front/src/utils/scanProjectBlockers.ts
import type { ScanProjectBlockers } from '@/types/scanProject'

export function formatScanProjectBlockersMessage(b: ScanProjectBlockers): string {
  const lines: string[] = [
    '本扫描项目已「添加为 Tool/Skill」并仍被以下 Agent 的 tools / skills 白名单引用。请先在「Agent 管理」中移除对应工具名或 Skill 名，再执行删除或重新扫描。',
    '',
  ]
  if (b.toolNames?.length) {
    lines.push(`· 工具名：${b.toolNames.join('、')}`)
  }
  if (b.skillNames?.length) {
    lines.push(`· Skill 名：${b.skillNames.join('、')}`)
  }
  if (b.agents?.length) {
    lines.push(`· 涉及 Agent：${b.agents.map((a) => a.name).join('、')}`)
  }
  return lines.join('\n')
}

/** 直接调删除/重扫返回 409 时，从 axios 错误中解析 body */
export function parseScanProjectBlockersFromError(err: unknown): ScanProjectBlockers | null {
  if (typeof err !== 'object' || err === null || !('response' in err)) {
    return null
  }
  const r = (err as { response?: { status?: number; data?: unknown } }).response
  if (r?.status !== 409 || r.data == null || typeof r.data !== 'object') {
    return null
  }
  const d = r.data as Record<string, unknown>
  if (typeof d.blocked !== 'boolean') {
    return null
  }
  return d as unknown as ScanProjectBlockers
}
